DRCluster mirrors S3 health reported via DRClusterConfig

raaizik opened the pull request:
### Changes
 - Adds s3 health condition to `DRClusterConfig` and reports in its status
 - DRCluster reconciler now reads DRClusterConfig status and sets S3Healthy condition:
      - If DRClusterConfigS3Healthy=True → S3Healthy=True, Reason=S3Reachable
      - Else → S3Healthy=False, Reason=S3Unreachable
      - Message mirrors DRClusterConfig condition message

**Logic:** During reconcile, fetch DRClusterConfig via MCV, inspect its S3Healthy condition, and update DRCluster status using `statusConditionSetAndUpdate`.

**Rationale:** Makes DRCluster status reflective of DRClusterConfig health for consistent S3 connectivity reporting.

- Unit test updates:
  * Assert S3Healthy condition instead of legacy reasons
  * Remove redundant config-change S3 tests; add single negative test for invalid S3Profile in DRCluster spec
  * Fix CIDR invalid test to update existing resource (avoid AlreadyExists error)

Part of the solution for #1641 

### TODO

- [ ] Debug secrets failure in e2e job

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added new DRCluster and DRClusterConfig S3 health condition state/reason constants (including separate reachable/unreachable outcomes).
  * Implemented DRClusterConfig S3 health evaluation by validating configured S3 store profiles and writing the resulting `S3Healthy` condition.

* **Bug Fixes**
  * Updated DRCluster S3 health to mirror DRClusterConfig `S3Healthy` instead of performing direct connectivity checks per S3 profile.

* **Tests**
  * Updated controller tests to drive and assert the new DRClusterConfig→DRCluster S3Healthy mirroring behavior, including related fencing scenarios.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

am-agrawa commented on api/v1alpha1/drclusterconfig_types.go:
```diff
@@ -33,7 +33,7 @@ type DRClusterConfigSpec struct {
 const (
 	DRClusterConfigConfigurationProcessed string = "Processed"
-	DRClusterConfigS3Reachable            string = "Reachable"
+	DRClusterConfigS3Healthy              string = "MetadadataStoresHealthy"
```
```suggestion
	DRClusterConfigS3Healthy              string = "MetaDataStoresHealthy"
```

ShyamsundarR commented on internal/controller/drclusterconfig_controller.go:
```diff
@@ -272,12 +288,71 @@ func (r *DRClusterConfigReconciler) processCreateOrUpdate(
 
+	if err := r.validateS3Profiles(ctx, log, drCConfig); err != nil {
+		log.Info("Reconcile error", "error", err)
+		setDRClusterConfigConfigurationProcessedCondition(&drCConfig.Status.Conditions, drCConfig.Generation,
```
This should call `setDRClusterConfigConfigurationProcessedCondition` and not update the processed condition.

I see that the function `validateS3Profiles` is updating the specific condition, so here we should not update or change the processed condition.

ShyamsundarR commented on internal/controller/drclusterconfig_controller.go:
```diff
@@ -272,12 +288,71 @@ func (r *DRClusterConfigReconciler) processCreateOrUpdate(
+		return ctrl.Result{Requeue: true}, err
+	}
+
 	setDRClusterConfigConfigurationProcessedCondition(&drCConfig.Status.Conditions, drCConfig.Generation,
```
Move this above checking the s3 rechability, such that processed condition is set to the right value independent of s3 validation.

ShyamsundarR commented on internal/controller/drclusterconfig_controller.go:
```diff
@@ -272,12 +288,71 @@ func (r *DRClusterConfigReconciler) processCreateOrUpdate(
+	// Fetch the ramen config resource
+	_, ramenConfig, err := ConfigMapGet(ctx, r.Client)
+	if err != nil {
+		return fmt.Errorf("config map get: %w", err)
```
Update error message like so (just going by other cases in the code):
`failed to get Ramen configmap` instead of `config map get`

ShyamsundarR commented on internal/controller/drclusterconfig_controller.go:
```diff
@@ -272,12 +288,71 @@ func (r *DRClusterConfigReconciler) processCreateOrUpdate(
+			if !k8serrors.IsNotFound(err) {
+				log.Info("Reconcile error", "error", err)
+
+				return fmt.Errorf("get: %w", err)
```
Same/similar error message update here as well.

ShyamsundarR replied:
Also set the `setDRClusterConfigS3ReachableCondition` to false?

ShyamsundarR commented on internal/controller/drclusterconfig_controller.go:
```diff
@@ -272,12 +288,71 @@ func (r *DRClusterConfigReconciler) processCreateOrUpdate(
+		}
+
+		// Profile does have a secret. Check if it has connectivity and record in status accordingly
+		if reason, err := s3ProfileValidate(ctx, r.Client, r.ObjectStoreGetter, profile.S3ProfileName, types.NamespacedName{
```
(nit/future) move `s3ProfileValidate` to a more common file than in drcluster_controller.go?

raaizik replied:
Yeah, but in follow-up PRs I'm going to get rid of `validateS3Profile` which calls it in drcluster controller, so it won't be used anywhere else but in drclusterconfig 💭

raaizik replied:
Done ✅

ShyamsundarR commented on internal/controller/drclusterconfig_controller.go:
```diff
@@ -272,12 +288,71 @@ func (r *DRClusterConfigReconciler) processCreateOrUpdate(
+		}.String(), r.Log); err != nil {
+			setDRClusterConfigS3ReachableCondition(&drCConfig.Status.Conditions, drCConfig.Generation, err.Error(),
+				metav1.ConditionFalse, reason)
+		} else {
```
(nit) avoid `else` return early as validation failed.

ShyamsundarR commented on internal/controller/drclusterconfig_controller.go:
```diff
@@ -272,12 +288,71 @@ func (r *DRClusterConfigReconciler) processCreateOrUpdate(
+			setDRClusterConfigS3ReachableCondition(&drCConfig.Status.Conditions, drCConfig.Generation,
+				fmt.Sprintf("S3 profile %q is healthy", profile.S3ProfileName), metav1.ConditionTrue, DRClusterConfigS3Reachable)
+
+			break
```
There maybe other profiles that are incorrect or not reachable, we would need to loop through all and then update the condition outside the loop to True, once all S3 stores are reachable.

ShyamsundarR commented on internal/controller/drclusterconfig_controller.go:
```diff
@@ -272,12 +288,71 @@ func (r *DRClusterConfigReconciler) processCreateOrUpdate(
+			setDRClusterConfigS3ReachableCondition(&drCConfig.Status.Conditions, drCConfig.Generation,
+				fmt.Sprintf("Found an unhealthy S3 profile %q for which there's no secret", profile.S3ProfileName),
+				metav1.ConditionFalse, DRClusterConfigS3Unreachable)
+		}
```
Return early, as the condition is already false.

ShyamsundarR reviewed (commented):
Mostly nits, other than the check for all S3 stores before returning True from the validation function.

abhijeet219 commented on internal/controller/drclusterconfig_controller.go:
```diff
@@ -275,9 +290,66 @@ func (r *DRClusterConfigReconciler) processCreateOrUpdate(
+				fmt.Sprintf("Found an unhealthy S3 profile %q for which there's no secret", profile.S3ProfileName),
+				metav1.ConditionFalse, DRClusterConfigS3Unreachable)
+
+			return fmt.Errorf("secrect not found: %w", err)
```
typo here

abhijeet219 commented on internal/controller/drclusterconfig_controller.go:
```diff
@@ -275,9 +290,66 @@ func (r *DRClusterConfigReconciler) processCreateOrUpdate(
+				setDRClusterConfigS3ReachableCondition(&drCConfig.Status.Conditions, drCConfig.Generation,
+					fmt.Sprintf("Found an unhealthy S3 profile %q for which there's no secret", profile.S3ProfileName),
+					metav1.ConditionFalse, DRClusterConfigS3Unreachable)
+
```
The function returns as soon as any profile fails validation, and will not validate remaining profiles. SHouldn't we check all profiles before returning?

rakeshgm commented:
looks good to me.

raaizik committed:
controllers: mirror S3 health from DRClusterConfig

- DRCluster reconciler now reads DRClusterConfig status and sets S3Healthy condition:
  * If DRClusterConfigS3Healthy=True → S3Healthy=True, Reason=S3Reachable
  * Else → S3Healthy=False, Reason=S3Unreachable
  * Message mirrors DRClusterConfig condition message

- Unit test updates:
  * Assert S3Healthy condition instead of legacy reasons
  * Remove redundant config-change S3 tests; add single negative test for invalid S3Profile in DRCluster spec
  * Fix CIDR invalid test to update existing resource (avoid AlreadyExists error)

Signed-off-by: raaizik <132667934+raaizik@users.noreply.github.com>
